Hsmtool: enable dumping output descriptors of onchain wallet

darosior opened the pull request:
This adds a new `dumponchaindescriptors` command to `hsmtool` which can:
- Dump 2 `xpub`-based descriptors (one for each scriptPubKey type we create)
    - Useful for external service (see https://github.com/ElementsProject/lightning/issues/4110 for more details)
- ~~Possibly dump 2 `xpriv`-based descriptors (one for each scriptPubKey type we create)~~ Removed: footgun with anchor outputs (thanks niftynei!)
    - ~~Useful for backing up / restoring the onchain wallet easily~~

TODO:
- [x] A functional test restoring a descriptor on `bitcoind` asserting we retrieve all the funds
- [x] Manpage and an extended doc on how to restore the onchain wallet using `bitcoind`
- [x] Maybe add a network parameter finally ?.. ~~EDIT: maybe output both main and test extended keys ?..~~ EDIT 2: added a `testnet` parameter

Fixes https://github.com/ElementsProject/lightning/issues/4110
Ping @jb55 : finally got to https://github.com/jb55/clightning-dumpkeys/issues/2 :-)

darosior commented:
Ok, added a test, documented (and rebased on https://github.com/ElementsProject/lightning/pull/4146 so that i can build locally)

darosior commented:
Travis failure is an unrelated timeout

darosior commented:
Rebased on master after #4146 merge, added a missing changelog.

niftynei commented on common/descriptor_checksum.c:
```diff
@@ -0,0 +1,84 @@
+
+
+int descriptor_checksum(const char *descriptor, int desc_size, char *checksum,
+			int checksum_capacity)
```
Further, since this only ever returns `0` or `8`, you could update this to return a boolean.

darosior replied:
> and remove the usage of the 0 as a null terminator in this function

Argh actually it's a terrible idea: the checksum *is* a string.

> The Right Way to do this would be to introduce a new struct desc_checksum e.g.
struct desc_checksum {
        u8 csum[8];
}

Actually the right way is using `char`s

niftynei commented on tools/hsmtool.c:
```diff
@@ -368,6 +371,55 @@ static int guess_to_remote(const char *address, struct node_id *node_id,
+	else
+		get_hsm_secret(&hsm_secret, hsm_secret_path);
+
+	/* We use m/0/0/k as the derivation tree for onchain funds. */
```
Needs reference. Where in the codebase do we do this?

niftynei replied:
One of these comments is wrong then 

>  	 *  - m/iH/0/k corresponds to the k'th keypair of the external chain of 
 	 * account number i of the HDW derived from master m. 

is not the same is.

>  We use m/0/0/k as the derivation tree for onchain funds. 

For the record, the intention here is that you'd update the code comment so other readers know where you're getting the info from, not just post a reply in github.

niftynei commented on tools/hsmtool.c:
```diff
@@ -368,6 +371,55 @@ static int guess_to_remote(const char *address, struct node_id *node_id,
+		salt++;
+		/* ..Which is used to derive m/ */
+	} while (bip32_key_from_seed(bip32_seed, sizeof(bip32_seed),
+				     /* An xpub can easily be converted to a tpub */
```
~No. This has potential to be *extremely* confusing. We need to either be told what chain to output it for or figure it out from the directory structure.~

updated later.

niftynei commented on tools/hsmtool.c:
```diff
@@ -368,6 +371,55 @@ static int guess_to_remote(const char *address, struct node_id *node_id,
+
+	/* The root seed is derived from hsm_secret using hkdf.. */
+	do {
+		hkdf_sha256(bip32_seed, sizeof(bip32_seed),
```
Duplicating this code from `hsmd/hsmd.c#populate_secretstuff` is a bad idea. Needs to be a shared routine that both hsmd and here use.

darosior replied:
That's how we've created tool since now ?..

niftynei replied:
I don't follow. We really want this code to be shared so that if one changes, the other also gets updated.

niftynei commented on tools/hsmtool.c:
```diff
@@ -368,6 +371,55 @@ static int guess_to_remote(const char *address, struct node_id *node_id,
+		errx(ERROR_LIBWALLY, "Can't encode xpub");
+
+	/* Now we format the descriptor strings (we only ever create P2WPKH and
+	 * P2SH-P2WPKH outputs). */
```
iirc, anchor outputs break this assumption

cdecker replied:
We could end up with anything if we use the `close_to` arguments, that includes addresses that we don't even control ourselves. The tool makes it clear we only show on-chain addresses that were generated by the internal wallet, so I don't think this'll be an issue.

darosior replied:
Yeah, now we don't promote it as a recovery mechanism anyways :-)

niftynei commented on tools/hsmtool.c:
```diff
@@ -417,6 +419,25 @@ static int dumponchaindescriptors(const char *hsm_secret_path, const char *passw
+		if (bip32_key_to_base58(&master_extkey, BIP32_FLAG_KEY_PRIVATE, &enc_xpub) != WALLY_OK)
+			errx(ERROR_LIBWALLY, "Can't encode xpub");
+
+		descriptor = tal_fmt(NULL, "wpkh(%s/0/0/*)", enc_xpub);
+		descriptor_checksum(descriptor, strlen(descriptor), checksum, sizeof(checksum));
+		printf("%s#%s\n", descriptor, checksum);
+		tal_free(descriptor);
```
this is the 3rd time you're repeating this code (and you do it again below).

darosior replied:
This commit is now removed.

niftynei commented on tools/hsmtool.c:
```diff
@@ -417,6 +419,25 @@ static int dumponchaindescriptors(const char *hsm_secret_path, const char *passw
+	/* If they also want xpriv break a line and output them as well. */
+	if (private) {
+		printf("\n");
+		if (bip32_key_to_base58(&master_extkey, BIP32_FLAG_KEY_PRIVATE, &enc_xpub) != WALLY_OK)
```
you're writing the `enc_xpriv` to a variable named `enc_xpub`

darosior replied:
This commit is now removed.

niftynei commented on tools/hsmtool.c:
```diff
@@ -373,11 +373,12 @@ static int guess_to_remote(const char *address, struct node_id *node_id,
 	struct secret hsm_secret;
 	u8 bip32_seed[BIP32_ENTROPY_LEN_256];
 	u32 salt = 0;
+	u32 version = testnet ? BIP32_VER_TEST_PRIVATE : BIP32_VER_MAIN_PRIVATE;
```
Can we infer this from the directory structure or some other info source?

darosior replied:
No, as we can use all commands of the `hsmtool` on a `hsm_secret` without depending on its environment at all
and this would be quite convoluted too..

cdecker replied:
We have all of these parameters in `chainparams.c`, so why not use them? We can even steal the `--network` option from `lightningd` itself :-)

darosior replied:
This'd mean bringing the `opt()` system here: that's a much bigger change :/

niftynei reviewed (changes requested):
Cool addition. I'm not sure how anchor outputs affects the assumption that all outputs are some variant of `p2wpkh`, may not be relevant for this patch.

darosior commented:
> I'm not sure how anchor outputs affects the assumption that all outputs are some variant of p2wpkh, may not be relevant for this patch.

You are right that we actually don't sweep the `1 CSV`-obstructed `to_remote`s in `onchaind` and just adverise them to the wallet, so i don't think i should actually sell this command as a "backup-and-recovery mechanism" in the FAQ..
I was also in doubt wrt dumping `xprivs` at all then, but it might turns to be useful (if i remove all mentions that it is exhaustive from the doc).

niftynei commented:
This functionality is Cool and Interesting, but until miniscript is widely adopted I think it's best to advertise this as a *subset* of your lightning funds tracking, at best. There's definitely "outputs we own, eventually" that haven't been swept to a 'recognizable wallet output' that will be missed by this -- if for some reason you take your lightning node down before these are swept there's a possibility that they'll be lost.

With these considerations in mind, I'd be tempted to remove the xpriv, with the rationale that it gives the false impression of 'total wallet funds control' when that's really not the case.

niftynei commented on tools/hsmtool.c:
```diff
@@ -469,7 +490,9 @@ int main(int argc, char *argv[])
-		return dumponchaindescriptors(argv[2], argc > 3 ? argv[3] : NULL);
+		return dumponchaindescriptors(argv[2],
+					      argc > 3 && !streq(argv[3], "") ? argv[3] : NULL,
+					      argc > 4 && streq(argv[4], "private"));
```
it's weird to me that you have to physically type in the phrase "private" and not look for a true/false here.

darosior replied:
I find it better: at usage typing `testnet` is way more meaningful than `true`

niftynei replied:
Update the parameter name then. As written, the parameter `testnet` suggests to say 'flag true if testnet'. 

Same thing for `private`.

niftynei commented on tools/hsmtool.c:
```diff
@@ -492,7 +491,8 @@ int main(int argc, char *argv[])
 					      argc > 3 && !streq(argv[3], "") ? argv[3] : NULL,
-					      argc > 4 && streq(argv[4], "private"));
+					      argc > 4 && streq(argv[4], "private"),
+					      argc > 5 && streq(argv[5], "testnet"));
```
same as previous comment, this seems like it should be checking for 'true/false' not "testnet"

niftynei reviewed (commented):
few more things I noticed while testing.

darosior commented on common/descriptor_checksum.h:
```diff
@@ -0,0 +1,11 @@
+/* https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md#reference */
+#define TOOL_DESCRIPTOR_CHECKSUM_LENGTH 8
+
+int descriptor_checksum(const char *descriptor, int desc_size, char *checksum,
```
Apart from taking bitcoind's implementation (which is virtually this) and asserting they are equel in the fuzz target, i don't see how there could so not specifically for the checksum result.
However the test added later on would fail to import the descriptor if the checksum was incorrect, so there is already that.

darosior commented on common/descriptor_checksum.c:
```diff
@@ -0,0 +1,84 @@
+	// Prevent appending zeroes from not affecting the checksum.
+	c ^= 1;
+
+	for (j = 0; j < 8; ++j)
```
Yeah i messed up my history (again...)

darosior commented:
- Used a proper API for the imported descriptor checksum file (use a struct instead and return a bool) as requested
- Addressed misc nits
- Removed the "output xprivs" functionality altogether, as @niftynei agrees it's a footgun. (We now use a watchonly wallet in the functional test).
- Rebased on master

cdecker commented on contrib/pyln-testing/pyln/testing/fixtures.py:
```diff
@@ -132,6 +132,14 @@ def bitcoind(directory, teardown_checks):
+    # FIXME: if we update the testsuite to use the upcoming 0.21 release we
+    # could switch to descriptor wallets and speed bitcoind operations
+    # consequently.
+    if not bitcoind.rpc.listwallets():
+        bitcoind.rpc.createwallet("lightningd-tests")
```
This is slightly nicer than 90f4ea62e856fdac3adf1b59235bb5f69eb609d1

cdecker commented on tests/test_wallet.py:
```diff
@@ -1045,6 +1045,36 @@ def test_hsmtool_secret_decryption(node_factory):
     assert node_id == l1.rpc.getinfo()["id"]
 
 
+@unittest.skipIf(TEST_NETWORK == 'liquid-regtest', '')
```
Add a comment saying that this is only supported for `mainnet` and `testnet` (technically we'd require it to be `regtest` but that appears to match `testnet`).

darosior replied:
Yeah descriptors for regtest are the same than for testnet (but addresses the same as mainnet!! :sob: )

cdecker replied:
Hurray for consistency 🤦

niftynei commented on tools/hsmtool.c:
```diff
@@ -33,6 +35,7 @@ static void show_usage(const char *progname)
 	       "<path/to/hsm_secret> [hsm_secret password]\n");
 	printf("	- guesstoremote <P2WPKH address> <node id> <tries> "
 	       "<path/to/hsm_secret> [hsm_secret password]\n");
+	printf("	- dumponchaindescriptors <path/to/hsm_secret> <password>\n");
```
We shouldn't be doing it for other commands. This is a security hole.

niftynei commented:
I opened an issue for the password inputs for all of the `hsmtool` commands, #4186. I think it's ok to merge this as is. We should address this issue though, as it's a potential security hole.

rustyrussell commented:
You should use ccan/opt, now that cmdline parsing is non-trivial.  But can be added after....

niftynei commented:
Updated the interface to use "network" instead of 'testnet', rebased on master now that #4065 is merged.

darosior commented:
ACK 29e576459967e963ceb2dbb4de463b04dc71f479

Thanks..
